sql/core/src/main/java/org/apache/spark/sql/sources/v2/TableProvider.java

```diff
@@ -25,7 +25,10 @@
  * The base interface for v2 data sources which don't have a real catalog. Implementations must
  * have a public, 0-arg constructor.
  * <p>
- * The major responsibility of this interface is to return a {@link Table} for read/write.
+ * The major responsibility of this interface is to return a {@link Table} for read/write. If you
+ * want to allow end-users to write data to non-existing tables via write APIs in `DataFrameWriter`
+ * with `SaveMode`, you must return a {@link Table} instance even if the table doesn't exist. The
+ * table schema can be empty in this case.
```

Comment:
What does it mean to write to a non-existing table? If you're writing somewhere, the table must exist.

This is for creating a table directly from configuration and an implementation class in the DataFrameWriter API. The target of the write still needs to exist.

Reply:
"Exist" is a relative concept, I suppose. I think we need to somehow allow for create-on-write functionality, even if many table providers won't want to support it.

Reply:
@jose-torres, create on write is done by CTAS. It should not be left up to the source whether to fail or create.

I think the confusion here is that this is a degenerate case where Spark has no ability to interact with the table's metadata. Spark must assume that it exists because the caller is writing to it.

The caller is indicating that a table exists, is identified by some configuration, and that a specific implementation can be used to write to it. That's what happens today when source implementations are directly specified.

Reply:
I don't want to break existing use cases, file sources can overwrite/append to a non-existing location, and we still need to support that with `SaveMode`.

Whatever the new write API will be, I think we still need to support `SaveMode` for a while.

Reply:
I think we can remove SaveMode right away. We don't need to break existing use cases if we add the OverwriteData plan and use it when the user's mode is Overwrite. That helps us get to the point where we can integrate SQL on top of this faster.

Reply:
I'm not convinced it's safe to remove `SaveMode` right away, when there is only an `Append` operator implemented currently.

If we do it, it means `DataFrameWriter.save` need to throw an exception for a lot of cases, except when the `mode` is append. I don't think this is acceptable right now.

Can we discuss the removal of `SaveMode` at least after all the necessary new write operators are implemented?

Reply:
`SaveMode` is incompatible with the SPIP to standarize behavior that was voted on and accepted. The save mode in `DataFrameWriter` must be used to create v2 plans that have well-defined behavior and cannot be passed to implementations in the final version of the v2 read/write API.

I see no reason to put off removing `SaveMode` from the API. If we remove it now, we will avoid having more versions of this API that are **fundamentally broken**. We will avoid more implementations that rely on it, not aware that it will be removed.

To your point about whether it is safe: the only case where this is actually used is `SaveMode.Overwrite` and `SaveMode.Append`. To replace those, all that needs to happen is to define what kind of overwrite should happen here (dynamic or truncate).

I can supply the logical plan and physical implementation in a follow-up PR because I already have all this written and waiting to go in. Or, I can add a PR to merge first if you'd like to have these changes depend on that implementation.

Reply:
what about the file source behavior difference between `SaveMode.Append` and the new append operator? Are you saying we should accept it and ask users to change their code? file source is widely used with `df.write.save` API...